### ISSUE TITLE
ValueError using addSolvent() with Charmm forcefield

### ISSUE THREAD

**Issue by rikebuck** [pandegroup/openmm#2139]
Hello, 
I am attempting to use the CHARMM force field on the pdb file attached ('1klu_fixed.pdb'), however get the following error at the addSolvent step: "ValueError: No template found for residue 180 (GLY).  The set of atoms matches ACEM, but the bonds are different.  Perhaps the chain is missing a terminal group?" What does this error mean/ any ideas to resolve it? Thank you.
```
pdb = PDBFile('1klu_fixed.pdb')

modeller = Modeller(pdb.topology, pdb.positions)

forcefield = ForceField('charmm36.xml', 'charmm36/water.xml')

modeller.addSolvent(forcefield, padding=2*nanometer, ionicStrength=0.1*molar, positiveIon='Na+', negativeIon = 'Cl-') 
```

[1klu_fixed.pdb.zip](https://github.com/pandegroup/openmm/files/2263107/1klu_fixed.pdb.zip)

**Comment by peastman** [comment]
@jchodera I think this might be another issue in the generation of the CHARMM force field.  The template for `GLY` does not list `NTER` as an allowed patch.  Hence it fails when a model has an N-terminal GLY.  Can you take a look and see why that's being left out?

**Comment by jchodera** [comment]
Looking into this!

**Comment by jchodera** [comment]
@peastman : Previously, `NTER` was not used to patch N-terminal `GLY`, but rather the `GLYP` patch:
```
    <Residue name="GLY">
      <Atom charge="-0.47" name="N" type="NH1"/>
      <Atom charge="0.31" name="HN" type="H"/>
      <Atom charge="-0.02" name="CA" type="CT2"/>
      <Atom charge="0.09" name="HA1" type="HB2"/>
      <Atom charge="0.09" name="HA2" type="HB2"/>
      <Atom charge="0.51" name="C" type="C"/>
      <Atom charge="-0.51" name="O" type="O"/>
      <Bond atomName1="N" atomName2="HN"/>
      <Bond atomName1="N" atomName2="CA"/>
      <Bond atomName1="C" atomName2="CA"/>
      <Bond atomName1="CA" atomName2="HA1"/>
      <Bond atomName1="CA" atomName2="HA2"/>
      <Bond atomName1="O" atomName2="C"/>
      <ExternalBond atomName="N"/>
      <ExternalBond atomName="C"/>
      <AllowPatch name="GLYP"/>
      <AllowPatch name="ACE"/>
      <AllowPatch name="ACED"/>
      <AllowPatch name="ACP"/>
      <AllowPatch name="ACPD"/>
      <AllowPatch name="NGNE"/>
      <AllowPatch name="CTER"/>
      <AllowPatch name="CNEU"/>
      <AllowPatch name="CTP"/>
      <AllowPatch name="CT2"/>
      <AllowPatch name="CT3"/>
    </Residue>
```
Our [last ParmEd PR eliminated duplicate patches](https://github.com/ParmEd/ParmEd/pull/978) by introducing an automated approach you suggested. (Diff for output appears in [this `openmm-forcefields` PR](https://github.com/choderalab/openmm-forcefields/pull/64)). This automated scheme we introduced appears to have [eliminated the `GLYP` patch](https://github.com/choderalab/openmm-forcefields/pull/64/files?utf8=%E2%9C%93&diff=split#diff-a80625b1448ccbcf8c977f83ad1631eeL93422) that patched N-terminal GLY residues.

Is there a simple way to patch the [logic for how we determine whether patches are equivalent](https://github.com/ParmEd/ParmEd/pull/978/files#diff-243f4c389f2827d4bb1db141b9ebe4c9R205)?

**Comment by peastman** [comment]
Why does it remove the GLYP patch?  What other patch does it think that's equivalent to?

**Comment by peastman** [comment]
> If we eliminate a patch Y as being identical to X, include X in the list of patches the residue is compatible with if it was compatible with Y

That sounds like the right thing to do.  We don't want to eliminate any patch/residue combinations.  We're just collapsing identical (as far as OpenMM is concerned) patches into a single definition.  That patch should still be applied to all residues the original redundant patches were applied to.

**Comment by jchodera** [comment]
On further thought, however, would that solution work in this case? 

Applying the `NTER` patch to `GLY` would give different charges than applying `GLYP` would for `CA` and `HA` atoms.

**Comment by peastman** [comment]
> The patch is not identical, but because the GLYP patch applied to a random compatible residue (presumably not a GLY) is identical to the NTER-patched residue to OpenMM, we remove it as being redundant.

It's not clear to me why it thinks those patches are identical?  GLYP involves atoms called HA1 and HA2.  GLY appears to be the only amino acid that has those, so it should be guaranteed that it *will* pick GLY to apply it to?  NTER on the other hand involves an atom called HA which doesn't exist in GLY, so it should be guaranteed to pick a different amino acid for that.  There doesn't exist any amino acid that is compatible with both GLYP and NTER.

The basic principle is that two patches are considered identical if the `<Patch>` tags are identical.  Looking at the tag for `NTER`, I see it includes a `<ChangeAtom>` for HA.  Since GLYP doesn't involve any atom called HA, its `<Patch>` tag will certainly not include that, and therefore won't be identical to NTER.

**Comment by jchodera** [comment]
@peastman : The logic for comparing equivalent patches [is here](https://github.com/ParmEd/ParmEd/blob/master/parmed/openmm/parameters.py#L205-L207). It looks like we only checked for equivalents of bonds to be added and atoms to be deleted. I'm not quite sure how I ended up implementing this limited check instead of the [more robust check for identity you recommended](https://github.com/pandegroup/openmm/issues/2047#issuecomment-382568295), but I'll get this fixed today.

**Comment by peastman** [comment]
Great, thanks!

**Comment by jchodera** [comment]
@peastman : I have a PR open to fix this, but I'm a bit worried now since *no* patches are being judged as equivalent:
https://github.com/ParmEd/ParmEd/pull/1000

I've attached the ffxml files that are built with this version in a ZIP archive.
[charmm36.zip](https://github.com/pandegroup/openmm/files/2268228/charmm36.zip)

**Comment by jchodera** [comment]
Correction: Only these patches were considered identical and removed:
```
UserWarning: Patch <PatchTemplate LIG1: 0 atoms; 0 bonds; head=None; tail=None> discarded because OpenMM considers it identical to <PatchTemplate LINK: 0 atoms; 0 bonds; head=None; tail=None>
UserWarning: Patch <PatchTemplate LIG2: 0 atoms; 0 bonds; head=None; tail=None> discarded because OpenMM considers it identical to <PatchTemplate LINK: 0 atoms; 0 bonds; head=None; tail=None>
UserWarning: Patch <PatchTemplate LIG3: 0 atoms; 0 bonds; head=None; tail=None> discarded because OpenMM considers it identical to <PatchTemplate LINK: 0 atoms; 0 bonds; head=None; tail=None>
UserWarning: Patch <PatchTemplate FHEM: 0 atoms; 0 bonds; head=None; tail=None> discarded because OpenMM considers it identical to <PatchTemplate DELB: 0 atoms; 0 bonds; head=None; tail=None>
```

**Comment by jchodera** [comment]
For example, you noted [in this post](https://github.com/pandegroup/openmm/issues/2047#issuecomment-382093395) that
> The ACE, ACED, ACP, and ACPD patches are all identical

When we're now ensuring that atom type and charge of added or changed atoms must be identical for two patches to be identical, we end up with separate patches for these residues:
```XML
 <Patch name="ACE">
      <AddAtom charge="-0.27" name="CAY" type="CT3"/>
      <AddAtom charge="0.09" name="HY1" type="HA3"/>
      <AddAtom charge="0.09" name="HY2" type="HA3"/>
      <AddAtom charge="0.09" name="HY3" type="HA3"/>
      <AddAtom charge="0.51" name="CY" type="C"/>
      <AddAtom charge="-0.51" name="OY" type="O"/>
      <AddBond atomName1="CY" atomName2="CAY"/>
      <AddBond atomName1="CY" atomName2="N"/>
      <AddBond atomName1="CAY" atomName2="HY1"/>
      <AddBond atomName1="CAY" atomName2="HY2"/>
      <AddBond atomName1="CAY" atomName2="HY3"/>
      <AddBond atomName1="OY" atomName2="CY"/>
      <RemoveExternalBond atomName="N"/>
    </Patch>
    <Patch name="ACED">
      <AddAtom charge="-0.27" name="CAY" type="CT3"/>
      <AddAtom charge="0.09" name="HY1" type="HA3"/>
      <AddAtom charge="0.09" name="HY2" type="HA3"/>
      <AddAtom charge="0.09" name="HY3" type="HA3"/>
      <AddAtom charge="0.51" name="CY" type="C"/>
      <AddAtom charge="-0.51" name="OY" type="O"/>
      <AddBond atomName1="CY" atomName2="CAY"/>
      <AddBond atomName1="CY" atomName2="N"/>
      <AddBond atomName1="CAY" atomName2="HY1"/>
      <AddBond atomName1="CAY" atomName2="HY2"/>
      <AddBond atomName1="CAY" atomName2="HY3"/>
      <AddBond atomName1="OY" atomName2="CY"/>
      <RemoveExternalBond atomName="N"/>
    </Patch>
    <Patch name="ACP">
      <AddAtom charge="-0.27" name="CAY" type="CT3"/>
      <AddAtom charge="0.09" name="HY1" type="HA3"/>
      <AddAtom charge="0.09" name="HY2" type="HA3"/>
      <AddAtom charge="0.09" name="HY3" type="HA3"/>
      <AddAtom charge="0.51" name="CY" type="C"/>
      <AddAtom charge="-0.51" name="OY" type="O"/>
      <AddBond atomName1="CY" atomName2="CAY"/>
      <AddBond atomName1="CY" atomName2="N"/>
      <AddBond atomName1="CAY" atomName2="HY1"/>
      <AddBond atomName1="CAY" atomName2="HY2"/>
      <AddBond atomName1="CAY" atomName2="HY3"/>
      <AddBond atomName1="OY" atomName2="CY"/>
      <RemoveExternalBond atomName="N"/>
    </Patch>
    <Patch name="ACPD">
      <AddAtom charge="-0.27" name="CAY" type="CT3"/>
      <AddAtom charge="0.09" name="HY1" type="HA3"/>
      <AddAtom charge="0.09" name="HY2" type="HA3"/>
      <AddAtom charge="0.09" name="HY3" type="HA3"/>
      <AddAtom charge="0.51" name="CY" type="C"/>
      <AddAtom charge="-0.51" name="OY" type="O"/>
      <AddBond atomName1="CY" atomName2="CAY"/>
      <AddBond atomName1="CY" atomName2="N"/>
      <AddBond atomName1="CAY" atomName2="HY1"/>
      <AddBond atomName1="CAY" atomName2="HY2"/>
      <AddBond atomName1="CAY" atomName2="HY3"/>
      <AddBond atomName1="OY" atomName2="CY"/>
      <RemoveExternalBond atomName="N"/>
    </Patch>
```
Their XML representations *are* identical, so something must be wrong with [my comparison routine](https://github.com/ParmEd/ParmEd/pull/1000/files#diff-243f4c389f2827d4bb1db141b9ebe4c9L205).

**Comment by jchodera** [comment]
Hm, I realize now there was *already* a scheme for eliminating duplicate patches using [this template hasher](https://github.com/ParmEd/ParmEd/blob/6d3fc9d1bb9af0383c28151de6f61da129b539e3/parmed/openmm/parameters.py#L486-L492) which was never extended to polyatomic patches.

I think the proper thing to do here is to instead update the logic of the [template hasher](https://github.com/ParmEd/ParmEd/blob/6d3fc9d1bb9af0383c28151de6f61da129b539e3/parmed/openmm/parameters.py#L486-L492) to give residues that will render to equivalent XML files identical hashes, and then use this to ensure all patches are unique when converting from a `CHARMMParameterSet` to an `OpenMMParameterSet`. I should be able to finish this tonight.

**Comment by jchodera** [comment]
I think I've managed to fix this by implementing a better [template hasher](https://github.com/ParmEd/ParmEd/pull/1000/files#diff-243f4c389f2827d4bb1db141b9ebe4c9R471). @peastman: Can you check that the information here is sufficient to include everything in residues and patches that OpenMM XML elements would contain?

The resulting code performs more reasonably on charmm36:
> 50 patches discarded, 197 retained

I've attached the updated ffxml files. Give this a try?
[charmm36.zip](https://github.com/pandegroup/openmm/files/2268837/charmm36.zip)

**Comment by peastman** [comment]
Thanks!  Looking through the changes I see it now has several additional patches, so that's good.  But I also see some problems:

- The DISU patch has been removed.  Did you somehow revert the changes that fixed #2039?
- A whole lot of residues have been removed.  Did the set of input files change?
- A lot of changes are showing up in the parameter definitions for the CustomTorsionForce.  Possibly it's just that they're in a different order from before, but it's hard to tell.

**Comment by jchodera** [comment]
> The DISU patch has been removed. Did you somehow revert the changes that fixed #2039?

Nope, but you [manually added the `DISU` patch](https://github.com/pandegroup/openmm/issues/2039#issuecomment-379115263) since fixing that issue requires major code changes we have not yet made.

> A whole lot of residues have been removed. Did the set of input files change?

My guess is that updating the template hasher resulted in better culling of identical residues, since this is used to [ensure only unique residues are written](https://github.com/ParmEd/ParmEd/blob/a7bb55d648c46d925d46ddc542356ea26ef9cf4e/parmed/openmm/parameters.py#L537-L549) as well. Indeed, looking at the output yields many messages like:
```
/Users/choderaj/miniconda/lib/python3.5/site-packages/ParmEd-3.0.3+47.gea04aa6.dirty-py3.5-macosx-10.9-x86_64.egg/parmed/openmm/parameters.py:548: UserWarning: Skipping writing of residue <ResidueTemplate CPES: 15 atoms; 15 bonds; head=None; tail=None> because OpenMM considers it identical to <ResidueTemplate CPEN: 15 atoms; 15 bonds; head=None; tail=None>
  warnings.warn('Skipping writing of residue {} because OpenMM considers it identical to {}'.format(residue, residue_collision))
/Users/choderaj/miniconda/lib/python3.5/site-packages/ParmEd-3.0.3+47.gea04aa6.dirty-py3.5-macosx-10.9-x86_64.egg/parmed/openmm/parameters.py:548: UserWarning: Skipping writing of residue <ResidueTemplate CNT1: 29 atoms; 31 bonds; head=None; tail=None> because OpenMM considers it identical to <ResidueTemplate CNTH: 29 atoms; 31 bonds; head=None; tail=None>
  warnings.warn('Skipping writing of residue {} because OpenMM considers it identical to {}'.format(residue, residue_collision))
/Users/choderaj/miniconda/lib/python3.5/site-packages/ParmEd-3.0.3+47.gea04aa6.dirty-py3.5-macosx-10.9-x86_64.egg/parmed/openmm/parameters.py:548: UserWarning: Skipping writing of residue <ResidueTemplate CNT2: 29 atoms; 31 bonds; head=None; tail=None> because OpenMM considers it identical to <ResidueTemplate CNTH: 29 atoms; 31 bonds; head=None; tail=None>
  warnings.warn('Skipping writing of residue {} because OpenMM considers it identical to {}'.format(residue, residue_collision))
/Users/choderaj/miniconda/lib/python3.5/site-packages/ParmEd-3.0.3+47.gea04aa6.dirty-py3.5-macosx-10.9-x86_64.egg/parmed/openmm/parameters.py:548: UserWarning: Skipping writing of residue <ResidueTemplate CNT3: 29 atoms; 31 bonds; head=None; tail=None> because OpenMM considers it identical to <ResidueTemplate CNTH: 29 atoms; 31 bonds; head=None; tail=None>
  warnings.warn('Skipping writing of residue {} because OpenMM considers it identical to {}'.format(residue, residue_collision))
/Users/choderaj/miniconda/lib/python3.5/site-packages/ParmEd-3.0.3+47.gea04aa6.dirty-py3.5-macosx-10.9-x86_64.egg/parmed/openmm/parameters.py:548: UserWarning: Skipping writing of residue <ResidueTemplate RM3P: 20 atoms; 20 bonds; head=None; tail=None> because OpenMM considers it identical to <ResidueTemplate TM3P: 20 atoms; 20 bonds; head=None; tail=None>
  warnings.warn('Skipping writing of residue {} because OpenMM considers it identical to {}'.format(residue, residue_collision))
/Users/choderaj/miniconda/lib/python3.5/site-packages/ParmEd-3.0.3+47.gea04aa6.dirty-py3.5-macosx-10.9-x86_64.egg/parmed/openmm/parameters.py:548: UserWarning: Skipping writing of residue <ResidueTemplate PIP2: 21 atoms; 21 bonds; head=None; tail=None> because OpenMM considers it identical to <ResidueTemplate PIP1: 21 atoms; 21 bonds; head=None; tail=None>
...
```
I've appended the complete output as a ZIP file for inspection.
[output.txt.zip](https://github.com/pandegroup/openmm/files/2271040/output.txt.zip)

> A lot of changes are showing up in the parameter definitions for the CustomTorsionForce. Possibly it's just that they're in a different order from before, but it's hard to tell.

I *thought* I had converted everything over from `dict` to `OrderedDict` to prevent things like this. Let me take a closer look.

**Comment by jchodera** [comment]
Looks like the `CustomTorsionForce` reordering might be due to the use of `dict` instead of `OrderedDict` in [this code](https://github.com/ParmEd/ParmEd/blob/a7bb55d648c46d925d46ddc542356ea26ef9cf4e/parmed/openmm/parameters.py#L339-L340). I'll fix that so we don't have that problem in future.

**Comment by jchodera** [comment]
Here's a tarball produced with the updated code, where ordering should subsequently remain stable.
[charmm36.zip](https://github.com/pandegroup/openmm/files/2271150/charmm36.zip)

**Comment by peastman** [comment]
Ok, I'll add back DISU by hand.

In the latest version, there are no longer changed parameters for CustomTorsionForce, just a whole lot of removed ones.  Possibly that's because of the removed residues?  But I don't see how that would happen.  It only removed residues because they were identical to other ones, so they should have exactly the same improper torsions.  Can you tell why lots of torsion definitions were removed?

Another difference is that before, a lot of residues supported the patch CY35, and that has been removed.  The output file reports it was removed because it was identical to LINK.  But there's no LINK patch defined in the file.  And the `<AllowPatch>` tags weren't changed to something else, they were just removed.

**Comment by jchodera** [comment]
> Can you tell why lots of torsion definitions were removed?

That's a good question. I'm seeing in `output.txt` a lot of messages like
```
/Users/choderaj/miniconda/lib/python3.5/site-packages/ParmEd-3.0.3+49.ga7bb55d.dirty-py3.5-macosx-10.9-x86_64.egg/parmed/openmm/parameters.py:424: ParameterWarning: No improper parameter found for ('NH1', 'CC', 'CT1', 'H')
```
which corresponds to the first improper that was dropped in `CustomTorsionForce`.

That [warning is issued by `_find_explicit_impropers()`](https://github.com/ParmEd/ParmEd/blob/master/parmed/openmm/parameters.py#L338-L424), which we added to explicitly enumerate all impropers needed by OpenMM since we ignore the CHARMM way of specifying impropers. It's possible that the logic here is affected by either our pruning of residues or patches or the changes to use `OrderedDict()` instead of `dict`. I'm looking into that now.

Another possibility is that [`_compress_impropers()`](https://github.com/ParmEd/ParmEd/blob/master/parmed/openmm/parameters.py#L429-L453) is eating some of the impropers. I'm adding a warning there to emit some text when this happens.

**Comment by jchodera** [comment]
I can verify that `_compress_impropers()` is not the culprit---it only compresses a couple of impropers.

The [`_find_explicit_impropers()` code](https://github.com/ParmEd/ParmEd/blob/master/parmed/openmm/parameters.py#L338-L424) looks more fishy to me, especially the alphabetical sorting.

Looking at the first warning:
```
ParameterWarning: No improper parameter found for ('NH1', 'CC', 'CT1', 'H')
```
Printing out the impropers in the parameter set shows that there are a few impropers that should match via wildcards:
```
(('CC', 'X', 'X', 'CT1'), <ImproperType; psi_k=96.000, psi_eq=0.000>)
(('NH1', 'X', 'X', 'H'), <ImproperType; psi_k=20.000, psi_eq=0.000>)
```
Presumably, [this code to check for wildcards](https://github.com/ParmEd/ParmEd/blob/master/parmed/openmm/parameters.py#L382-L402) should handle that case. I'll have to read through carefully to see why it might not be.

**Comment by jchodera** [comment]
@swails: I'm pretty sure the [`_find_explicit_impropers()` code](https://github.com/ParmEd/ParmEd/blob/master/parmed/openmm/parameters.py#L338-L424) will need to be rewritten, but I'm still trying to figure out what it was originally *intended* to do. Was the idea that we should just consider each improper torsion in each residue template and make sure that there is at least one OpenMM explicit improper definition that OpenMM will match, using some standard bonded atom ordering for impropers?

I'm not entirely sure why the behavior changed, but it appears that the code might make some assumptions about how the parameter set defines atom order in storing `improper_types` that may have changed in the interim.

**Comment by jchodera** [comment]
@peastman : I've finally had time to overhaul the CHARMM generic improper expansion.
* Updated ParmEd code: https://github.com/ParmEd/ParmEd/pull/1001
* Updated parameters: https://github.com/choderalab/openmm-forcefields/pull/64

I've attached the updated parameter set as a ZIP for easy inspection.

While this fixes the missing impropers, note that this sometimes leads to slightly different parameters from before. My current understanding is that this is either because the logic used to expand generics does not result in unique matches, or because the previous code (which was difficult to understand) was somewhat buggy.
[charmm36.zip](https://github.com/pandegroup/openmm/files/2280623/charmm36.zip)

**Comment by jchodera** [comment]
@peastman : If we're planning to release the updated parameters as a bugfix release for OpenMM 7.2, we should update the AMBER parameters to match the recent AmberTools 18 release before we release OpenMM 7.3.

**Comment by peastman** [comment]
Thanks, I'll take a look at it.

> If we're planning to release the updated parameters as a bugfix release for OpenMM 7.2, we should update the AMBER parameters to match the recent AmberTools 18 release before we release OpenMM 7.3.

I think we can just include it in 7.3.  If there are other changes you think should be in 7.3, now is the time to update them.

**Comment by jchodera** [comment]
> If there are other changes you think should be in 7.3, now is the time to update them.

Great---I'll update AmberTools now then.

**Comment by peastman** [comment]
@rikebuck an updated version of CHARMM36 that should fix this problem is in #2149.  Can you verify that it works for you?  Thanks!

**Comment by peastman** [comment]
Fixed in #2149.
